pilot/pkg/networking/core/v1alpha3/extension/wasmplugin.go

```diff
@@ -0,0 +1,149 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package extension
+
+import (
+	udpa "github.com/cncf/udpa/go/udpa/type/v1"
+	envoy_config_core_v3 "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	hcm_filter "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/network/http_connection_manager/v3"
+	"github.com/envoyproxy/go-control-plane/pkg/conversion"
+	"google.golang.org/protobuf/proto"
+
+	extensions "istio.io/api/extensions/v1alpha1"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/networking"
+	"istio.io/istio/pilot/pkg/networking/util"
+	authzmodel "istio.io/istio/pilot/pkg/security/authz/model"
+	securitymodel "istio.io/istio/pilot/pkg/security/model"
+)
+
+const (
+	wasmFilterType = "envoy.extensions.filters.http.wasm.v3.Wasm"
+)
+
+var defaultConfigSource = &envoy_config_core_v3.ConfigSource{
+	ConfigSourceSpecifier: &envoy_config_core_v3.ConfigSource_Ads{
+		Ads: &envoy_config_core_v3.AggregatedConfigSource{},
+	},
+	ResourceApiVersion: envoy_config_core_v3.ApiVersion_V3,
+}
+
+// AddWasmPluginsToMutableObjects adds WasmPlugins to HTTP filterChains
+// Note that the slices in the map must already be ordered by plugin
+// priority! This will be the case for maps returned by PushContext.WasmPlugin()
+func AddWasmPluginsToMutableObjects(
+	mutable *networking.MutableObjects,
+	extensionsMap map[extensions.PluginPhase][]*model.WasmPluginWrapper,
+) {
+	if mutable == nil {
+		return
+	}
+
+	for fcIndex, fc := range mutable.FilterChains {
+		// we currently only support HTTP
+		if fc.ListenerProtocol != networking.ListenerProtocolHTTP {
+			continue
+		}
+		mutable.FilterChains[fcIndex].HTTP = injectExtensions(fc.HTTP, extensionsMap)
+	}
+}
+
+func injectExtensions(filterChain []*hcm_filter.HttpFilter, exts map[extensions.PluginPhase][]*model.WasmPluginWrapper) []*hcm_filter.HttpFilter {
+	// copy map as we'll manipulate it in the loop
+	extMap := make(map[extensions.PluginPhase][]*model.WasmPluginWrapper)
+	for phase, list := range exts {
+		extMap[phase] = []*model.WasmPluginWrapper{}
+		extMap[phase] = append(extMap[phase], list...)
+	}
+	newHTTPFilters := make([]*hcm_filter.HttpFilter, 0)
+	for _, httpFilter := range filterChain {
```

Comment:
This part of logic feels a bit hard for me to follow.. Can we go with something simple like:

```go
// construct a first class extation map based on name.
firstClassExts := map[string]*hcm_filter.HttpFilter
for _, f := range filterChain {
  firstClassExts[f.Name] = f
}

// Construct a new ext list based on phase.
newHTTPFilters := make([]*hcm_filter.HttpFilter, 0)
newHTTPFilters = popAppend(newHTTPFilters, extMap, extensions.PluginPhase_AUTHN)
if f, ok := firstClassExts[securitymodel.EnvoyJwtFilterName]; ok {
  newHTTPFilters = append(newHTTPFilters, f)
}
if f, ok := firstClassExts[securitymodel.AuthnFilterName]; ok {
  newHTTPFilters = append(newHTTPFilters, f)
}
newHTTPFilters = popAppend(newHTTPFilters, extMap, extensions.PluginPhase_AUTHZ)
if f, ok := firstClassExts[firstClassExts[securitymodel.RBACFilterName]; ok {
  newHTTPFilters = append(newHTTPFilters, f)
}
newHTTPFilters = popAppend(newHTTPFilters, extMap, extensions.PluginPhase_STATS)
```

Not sure if anything would make this logic not work, but this feels easier to follow and has less duplications.

Reply:
I think we know all the first class plugins? They are a limited set and can be enumerated here.

Reply:
that's right- but what I mean is that this is incompatible with pilot's plugin system. There could be any number of filters in the chain that we don't know about depending on the plugins being used

Reply:
Not sure I follow. What is the incompatibility? Can you provide an example?